src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -566,49 +575,138 @@ LRESULT CALLBACK FancyZones::s_WndProc(HWND window, UINT message, WPARAM wparam,
+        device newDevice{{}, monitor};
+        if (displayDevice.DeviceID[0] != L'\0')
+        {
+            std::array<wchar_t, 256> parsedId{};
+            ParseDeviceId(displayDevice.DeviceID, parsedId.data(), size(parsedId));
+            newDevice._deviceID = parsedId.data();
+        }
+        devices->emplace_back(std::move(newDevice));
+        return doContinueEnumeration;
+    };
+    std::vector<device> devices;
+    EnumDisplayMonitors(nullptr, nullptr, callback, reinterpret_cast<LPARAM>(&devices));
+    std::vector<WmiMonitorID> monitor_infos;
+    try
+    {
+        m_wmiConnection.select_all(L"WmiMonitorID", [&](std::wstring_view xml_obj) {
+          monitor_infos.emplace_back(parse_monitorID_from_dtd(xml_obj));
+        });
+    }
+    catch(...) {}
 
-            if (validMonitor)
+    if(GetSystemMetrics(SM_REMOTESESSION))
+    {
+        // When running in a remote session, devices have "Default_Monitor#<transient_part>" DeviceIDS.
+        // Since a remote session can have multiple devices, we can't just assign them a single "REMOTE" ID.
+        // We'll assign them unique numeric IDs instead.
+        for (size_t i = 0; i < size(devices); ++i)
+        {
+            wchar_t remoteDeviceID[32];
+            swprintf_s(remoteDeviceID, L"REMOTE%zu", i);
+            AddZoneWindow(devices[i]._handle, remoteDeviceID);
+        }
+        return;
+    }
+    // If EnumDisplayMonitors returned a duplicate data, it's an API bug and we should workaround it
+    const bool enumDisplayMonitorsDublicateData = [&]{
+        std::unordered_set<std::wstring_view> discoveredDeviceIDs;
+        for (const auto & dev : devices)
+        {
+            const auto generatedPartOfID = dev.get_generated_part_of_id();
+            if (!generatedPartOfID.has_value())
             {
-                if (!deviceId)
-                {
-                    deviceId = GetSystemMetrics(SM_REMOTESESSION) ?
-                        L"\\\\?\\DISPLAY#REMOTEDISPLAY#" :
-                        L"\\\\?\\DISPLAY#LOCALDISPLAY#";
-                }
-
-                auto strongThis = reinterpret_cast<FancyZones*>(data);
-                strongThis->AddZoneWindow(monitor, deviceId);
+                continue;
+            }
+            const auto [_, hasUniqueID] = discoveredDeviceIDs.emplace(*generatedPartOfID);
+            if (!hasUniqueID)
+            {
+                return true;
             }
         }
-        return TRUE;
-    };
+        return false;
+    }();
 
+
+    for (size_t i = 0; i < size(devices); ++i)
     {
-        std::unique_lock writeLock(m_lock);
-        m_zoneWindowMap.clear();
+        const auto & dev = devices[i];
+        const auto generatedPartOfID = dev.get_generated_part_of_id();
+        // In case DeviceID doesn't contain valid data or we can't rely on it, since EnumDisplayMonitors
+        // returned a duplicate results, we try using WMI data or fallback to UNKNOWN + numeric ID.

```

Comment:
Change 
```
        // In case DeviceID doesn't contain valid data or we can't rely on it, since EnumDisplayMonitors
        // returned a duplicate results, we try using WMI data or fallback to UNKNOWN + numeric ID.
```

to

```
        // In case DeviceID doesn't contain valid data or EnumDisplayMonitors returned duplicate results,
        // try using WMI data, if WMI can't be used fallback to UNKNOWN + numeric ID.
```